client/async.go

```diff
@@ -0,0 +1,268 @@
+	a.config.SetDefault(numRoutinesConf, 5)
+	numSendRoutines := a.config.GetInt(numRoutinesConf)
+
+	a.logger = a.logger.WithFields(logrus.Fields{
+		"numRoutines": numSendRoutines,
+	})
+
+	for i := 0; i < numSendRoutines; i++ {
+		go a.sendRoutine()
+	}
+
+	return a, nil
+}
+
+func (a *gRPCClientAsync) configureGRPCForwarderClient(
+	serverAddress string,
+	client pb.GRPCForwarderClient,
+	opts ...grpc.DialOption,
+) error {
+	if client != nil {
+		a.client = client
+		return nil
+	}
+	a.logger.WithFields(logrus.Fields{
+		"operation": "configureGRPCForwarderClient",
+	}).Info("connecting to grpc server")
+	dialOpts := append(
+		[]grpc.DialOption{
+			grpc.WithInsecure(),
+			grpc.WithUnaryInterceptor(a.metricsReporterInterceptor),
+		},
+		opts...,
+	)
+	conn, err := grpc.Dial(serverAddress, dialOpts...)
+	if err != nil {
+		return err
+	}
+	a.conn = conn
+	a.client = pb.NewGRPCForwarderClient(conn)
+	return nil
+}
+
+// metricsReporterInterceptor will report metrics from client
+func (a *gRPCClientAsync) metricsReporterInterceptor(
+	ctx context.Context,
+	method string,
+	req interface{},
+	reply interface{},
+	cc *grpc.ClientConn,
+	invoker grpc.UnaryInvoker,
+	opts ...grpc.CallOption,
+) error {
+	l := a.logger.WithFields(logrus.Fields{
+		"method": method,
+	})
+
+	events := req.(*pb.SendEventsRequest).Events
+
+	defer func(startTime time.Time) {
+		timeUsed := float64(time.Since(startTime).Nanoseconds() / 1000000)
+		for _, e := range events {
+			metrics.ClientRequestsResponseTime.WithLabelValues(
+				hostname,
+				method,
+				e.Topic,
+			).Observe(timeUsed)
+		}
+		l.WithFields(logrus.Fields{
+			"timeUsed": timeUsed,
+			"reply":    reply.(*pb.SendEventsResponse),
+		}).Debug("request processed")
+	}(time.Now())
+
+	err := invoker(ctx, method, req, reply, cc, opts...)
+	if err != nil {
+		l.WithError(err).Error("error processing request")
+		for _, e := range events {
+			metrics.ClientRequestsFailureCounter.WithLabelValues(
+				hostname,
+				method,
+				e.Topic,
+				err.Error(),
+			).Inc()
+		}
+	} else {
+		for _, e := range events {
+			metrics.ClientRequestsSuccessCounter.WithLabelValues(
+				hostname,
+				method,
+				e.Topic,
+			).Inc()
+		}
+	}
+	return err
+}
+
+func (a *gRPCClientAsync) send(ctx context.Context, event *pb.Event) error {
+	atomic.AddUint64(&a.transientCounter, 1)
+	a.eventsChannel <- event
+	return nil
+}
+
+func (a *gRPCClientAsync) sendRoutine() {
+	ticker := time.NewTicker(a.flushInterval)
+	defer ticker.Stop()
+
+	req := &pb.SendEventsRequest{}
+	req.Events = make([]*pb.Event, 0, a.flushSize)
+
+	send := func() {
+		cpy := req
+		cpy.Id = uuid.NewV4().String()
+		req = &pb.SendEventsRequest{}
+		req.Events = make([]*pb.Event, 0, a.flushSize)
+		go a.sendEvents(cpy, 0)
+	}
+
+	for {
+		select {
+		case e := <-a.eventsChannel:
+			if len(req.Events) == 0 {
+				a.wg.Add(1)
+			}
+			atomic.AddUint64(&a.transientCounter, ^uint64(0))
```

Comment:
Since you want to decrement, why not use an int64? Do you really expect to accumulate that many events?